### PR TITLE
RATEST-298: Fix build failures in Appointment management workflow

### PR DESCRIPTION
**Pull request by Irenyak1**
Ticket ID -> [RATEST-298](https://issues.openmrs.org/browse/RATEST-298)

Description ->Added waiter method to solve build failures in Appointment management workflow

### CODE REVIEW

File: .github/workflows/refapp-2x-appointment.yml [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+      - name: run db and web containers
+        run: |
+          docker-compose -f docker/docker-compose-refqa.yml up -d
+      - name: wait for openmrs instance to start
+        run: while [[ "$(curl -s -o /dev/null -w ''%{http_code}'' http://localhost:8080/openmrs/login.htm)" != "200" ]]; do sleep 1; done
+      - name: Run qaframework on selenium
+        run: |
+          echo "cucumber.publish.enabled=true" > src/test/resources/cucumber.properties
+          mvn clean install -DskipTests=true
+          cp -f src/test/resources/org/openmrs/uitestframework/test-local.properties src/test/resources/org/openmrs/uitestframework/test.properties
```
Comment (mherman22) [inline .github/workflows/refapp-2x-appointment.yml:43]: does this work without first cding into qaframeworkbdd? i.e `cd qaframework-bdd-tests`

File: qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/AppointmentBlocksPage.java [hunk 2]
```diff
@@ -19,6 +19,7 @@
 public class AppointmentBlocksPage extends Page {
 
 	public static final By DELETE = By.linkText("Delete");
+	public static final By CURRENT_DAY_BUTTON = By.cssSelector("#calendar table tbody tr td.fc-header-left span.fc-button.fc-button-basicDay.fc-state-default.fc-corner-left.fc-corner-right");
```
Comment (sherrif10) [inline qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/AppointmentBlocksPage.java:22]: Can we use a different element , probably xpath. this is likely not convinient.

Comment (Irenyak1) [reply qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/AppointmentBlocksPage.java:22]: @mherman22 I don't know if this answers your question. I think line 8 does that.

Comment (mherman22) [reply qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/AppointmentBlocksPage.java:22]: > @mherman22 I don't know if this answers your question. I think line 8 does that.

yeah after reading through i saw it

Comment (Irenyak1) [reply qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/AppointmentBlocksPage.java:22]: > Can we use a different element , probably xpath. this is likely not convinient.

@sherrif10  earlier on like a few months back, I used xpath and this was causing the test to fail until I switched to css selector. However, let me give it another try. I will give you feedback.

Comment (Irenyak1) [reply qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/AppointmentBlocksPage.java]: @sherrif10  I have attended to the review and pushed a commit. You could take some time to make more reviews.

File: qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/AppointmentBlocksPage.java [hunk 3]
```diff
@@ -19,6 +19,7 @@
 public class AppointmentBlocksPage extends Page {
 
 	public static final By DELETE = By.linkText("Delete");
+	public static final By CURRENT_DAY_BUTTON = By.xpath("//*[@id=\"calendar\"]/table/tbody/tr/td[1]/span[5]");
```
Comment (kdaud) [inline qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/AppointmentBlocksPage.java:22]: How about using another locator rather than xpath?

File: qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/ManageAppointmentsPage.java [hunk 4]
```diff
@@ -14,38 +14,38 @@
...
-	private static final By SAVE = By.xpath("//div[@id='confirmAppointment']/div[2]/button[2]");
-	private static final By BOOK_APPOINTMENT = By.cssSelector("i.icon-calendar:nth-child(1)");
-	private static final By YES_DELETE_REQUEST = By.xpath("//div[@id='confirm-cancel-appointment-request']/div[2]/button");
-	private static final By VIEW_ALL_TYPES_LINK = By.cssSelector("#viewAllAppointmentTypes > a");
+	private static final By SAVE_BUTTON = By.xpath("//div[@id='confirmAppointment']/div[2]/button[2]");
+	private static final By CANCEL_BUTTON = By.cssSelector("#searchButtons button");
+	private static final By SEARCH_BUTTON = By.cssSelector("#searchButtons button.confirm");
+	private static final By NEXT_BUTTON = By.cssSelector("#selectAppointment button.confirm");
+	private static final By BOOK_APPOINTMENT_ICON = By.cssSelector("i.icon-calendar:nth-child(1)");
+	private static final By CONFIRM_CANCEL_APPOINTMENT_REQUEST_BUTTON  = By.xpath("//div[@id='confirm-cancel-appointment-request']/div[2]/button");
```
Comment (kdaud) [inline qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/ManageAppointmentsPage.java:24]: and here

File: qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/AppointmentBlocksPage.java (hunk 3 above)
Comment (Irenyak1) [reply qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/AppointmentBlocksPage.java:22]: > How about using another locator rather than xpath?

@kdaud  that button has a generic class name and it has not ID. I had initially used css selector and @sherrif10 advised me to use an xpath. I will continue looking into it to see the best selector I could come up with. Thanks.

File: qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/ManageAppointmentsPage.java (hunk 4 above)
Comment (Irenyak1) [reply qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/ManageAppointmentsPage.java:24]: Sure

File: qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/AppointmentBlocksPage.java (hunk 2 above)
Comment (kdaud) [reply qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/AppointmentBlocksPage.java]: I am sorry being late in the conversation here.

> Can we use a different element , probably xpath. this is likely not convinient

@sherrif10 I find the css locator being more reliable than using xpath so I would go with the locator suggested by @Irenyak1

Comment (sherrif10) [reply qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/AppointmentBlocksPage.java]: No objection still, as long as it won't break in future

File: qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/AppointmentBlocksPage.java (hunk 3 above)
Comment (kdaud) [reply qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/AppointmentBlocksPage.java:22]: @Irenyak1 checkout the comment [here](https://github.com/openmrs/openmrs-contrib-qaframework/pull/425#discussion_r899801884)

Comment (Irenyak1) [reply qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/AppointmentBlocksPage.java:22]: Sure, let me revert

File: qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/ManageAppointmentsPage.java (hunk 4 above)
Comment (Irenyak1) [reply qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/ManageAppointmentsPage.java]: @kdaud  I have attended to the reviews and pushed a commit,  please you could make more.

**Comment by Irenyak1** [comment]
@kdaud , @sherrif10, Are there any more reviews on this issue or it can be merged into master?

File: qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/ManageAppointmentsPage.java [hunk 5]
```diff
@@ -14,38 +14,38 @@
...
 public class ManageAppointmentsPage extends Page {
 
-	public static final By DELETE_REQUEST = By.cssSelector(".delete-item.icon-remove");
-	private static final By CANCEL_BUTTON = By.cssSelector("#searchButtons > button");
-	private static final By SEARCH = By.cssSelector("#searchButtons > button.confirm");
+	public static final By DELETE_REQUEST_ICON = By.cssSelector(".delete-item.icon-remove");
 	private static final By APPOINTMENT = By.xpath("//table[@id='appointmentTable']/div[2]/div/div/div/div[2]/div[2]/div");
-	private static final By NEXT = By.cssSelector("#selectAppointment > button.confirm");
-	private static final By SAVE = By.xpath("//div[@id='confirmAppointment']/div[2]/button[2]");
-	private static final By BOOK_APPOINTMENT = By.cssSelector("i.icon-calendar:nth-child(1)");
```
Comment (sherrif10) [inline qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/ManageAppointmentsPage.java:23]: Are these changes necessary.

Comment (Irenyak1) [reply qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/ManageAppointmentsPage.java:23]: @sherrif10 yes all those files are involved in the appointment tests. The changes were made such that the variable names make more sense depending on what they are standing for.

Comment (Irenyak1) [reply qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/ManageAppointmentsPage.java:23]: Are there more reviews on this?

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/AppointmentSteps.java [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+	public void userClicksOnDeleteAppointmentRequest() {
+		manageAppointmentsPage.deleteRequest() ;
+	}
+
+	@Then("the system deletes appointment request from the request table")
+	public void systemDeletesAppointmentRequest() {
+		assertTrue(manageAppointmentsPage.containsText("No appointment requests"));
+	}
+
+	@When("a user clicks Appointment scheduling app from the home page")
```
Comment (sherrif10) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/AppointmentSteps.java:108]:
```suggestion
	@When("a user clicks  on Appointment scheduling app from the home page")
```

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/AppointmentSteps.java [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+		manageAppointmentsPage.searchAppointment();
+		manageAppointmentsPage.clickAppointment();
+	}
+
+	@And("a user clicks on book appointment icon")
+	public void clickOnBookAppointment() {
+		manageAppointmentsPage.clickOnBookAppointment();
+	}
+
+	@And("a user clicks the Save button")
```
Comment (sherrif10) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/AppointmentSteps.java:236]:
```suggestion
	@And("a user clicks on  Save button")
```

Comment (sherrif10) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/AppointmentSteps.java:236]: @Irenyak1  Well done on this, Left few simply changes to consider, then this is good to be merged.

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/AppointmentSteps.java (hunk 6 above)
Comment (Irenyak1) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/AppointmentSteps.java:108]: @sherrif10 If I am to implement this change, the test will not run because the proposed change already exists in the VitalsAndTriagingSteps.

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/AppointmentSteps.java [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+	public void captureAppointmentRequestDetails() {
+		requestAppointmentPage.enterAppointmentType(FIRST_SERVICE_NAME);
+		requestAppointmentPage.enterMinimumValue("0");
+		requestAppointmentPage.selectMinimumUnits("Day(s)");
+		requestAppointmentPage.enterMaximumValue("2");
+		requestAppointmentPage.selectMaximumUnits("Day(s)");
+		requestAppointmentPage.addAppointmentRequestNote("I need to see a doctor first");
+	}
+
+	@And("a user clicks save button")
```
[the remark this answers is not in the document]
Comment (Irenyak1) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/AppointmentSteps.java:78]: @sherrif10 I did not put the word on because there's a workflow that used that so it throws a conflict if I include it.

**Comment by kdaud** [comment]
@Irenyak1 did you consent to have this test as one or you think of breaking it down into two workflows?

**Comment by kdaud** [comment]
Is there any new change you have made to fix the test which wasn't done in the previous two attempts when allowing this guy into master?

**Comment by Irenyak1** [comment]
> Is there any new change you have made to fix the test which wasn't done in the previous two attempts when this guy was allowed to enter into master?

Yes @kdaud  I changed a few things just like in[ this commit ](https://github.com/openmrs/openmrs-contrib-qaframework/pull/425/commits/29e2643b22e9cf16c5327f0928506c9bbbf0cf63#)

**Review by kdaud — APPROVED** [review]

**Comment by kdaud** [comment]
CI is unhappy with passing this test into it's territory: https://ci.openmrs.org/browse/CONTRIB-QA-1584

**Comment by kdaud** [comment]
@Irenyak1 you may need to think of the second alternative mentioned [here](https://github.com/openmrs/openmrs-contrib-qaframework/pull/425#issuecomment-1162146750)

**Comment by kdaud** [comment]
Taking a look at the cucumber report at https://reports.cucumber.io/reports/7228fb2b-1f7c-4031-9fac-36261f55a918 would be a good starting point to know the exact step failing in CI.

**Comment by sherrif10** [comment]
@Irenyak1  you can re use what already exist.

**Comment by kdaud** [comment]
@Irenyak1 do you have some updates on this ticket?

**Comment by Irenyak1** [comment]
Hello @kdaud , @sherrif10 , @mherman22 , @Muta-Jonathan  I have created [a new PR](https://github.com/openmrs/openmrs-contrib-qaframework/pull/441)  to handle what we were discussing here.
